plugins/auth-backend-module-pinniped-provider/src/authenticator.ts

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { PassportDoneCallback } from '@backstage/plugin-auth-node';
+import {
+  createOAuthAuthenticator,
+  decodeOAuthState,
+  encodeOAuthState,
+} from '@backstage/plugin-auth-node';
+import {
+  Client,
+  Issuer,
+  TokenSet,
+  Strategy as OidcStrategy,
+} from 'openid-client';
+
+const rfc8693TokenExchange = async ({
+  subject_token,
+  target_audience,
+  ctx,
+}: {
+  subject_token: string;
+  target_audience: string;
+  ctx: Promise<{
+    providerStrategy: OidcStrategy<{}>;
+    client: Client;
+  }>;
+}): Promise<string | undefined> => {
+  const { client } = await ctx;
+  return client
+    .grant({
+      grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
+      subject_token,
+      audience: target_audience,
+      subject_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+      requested_token_type: 'urn:ietf:params:oauth:token-type:jwt',
+    })
+    .then(tokenset => tokenset.access_token)
+    .catch(err => {
+      throw new Error(`RFC8693 token exchange failed with error: ${err}`);
+    });
+};
+
+/** @public */
+export const pinnipedAuthenticator = createOAuthAuthenticator({
+  defaultProfileTransform: async (_r, _c) => ({ profile: {} }),
+  async initialize({ callbackUrl, config }) {
+    const issuer = await Issuer.discover(
+      `${config.getString(
+        'federationDomain',
+      )}/.well-known/openid-configuration`,
+    );
```

Comment:
I'm feeling it would be quite a lot better to not do this upfront, but on-demand when needed and then cached for a while if successful. Right now this provider will be broken until the backend restarts if this initial request fails. The context created by the `initialize` method is really intended to not be async for that reason.

Reply:
I believe i would need to explore this a bit more and might even need some guidance if we want to move the issuer declaration outside of the `initialize` method. the current `OidcStrategy` that is used in this implementation depends on a client which depends on that issuer and I don't quite understand just yet where to call/cache the issuer's async call so that we can later use it in our initialize.